# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: toolkit/components/antitracking/ContentBlocking.cpp
# Commit: 3b1dbe909aaf
# Full Hash: 3b1dbe909aaf1982c72db1e5bc8cb35208ae6c77
# Author: Dimi Lee <dlee@mozilla.com>
# Date: 2020-04-23 21:43:09
# Regressor Bug: 1616775
# File Overlap Count: 1
# Description:
#   Bug 1616775 - P4. Add CompleteAllowAccessFor IPDL r=timhuang,baku
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D71013
# ==============================================================================

diff -r 123b6c3a0ede -r 3b1dbe909aaf toolkit/components/antitracking/ContentBlocking.cpp
--- a/toolkit/components/antitracking/ContentBlocking.cpp	Thu Apr 23 14:28:41 2020 +0000
+++ b/toolkit/components/antitracking/ContentBlocking.cpp	Thu Apr 23 14:25:16 2020 +0000
@@ -331,7 +331,29 @@
         behavior, aReason, aPerformFinalChecks);
   }
 
-  return StorageAccessGrantPromise::CreateAndReject(false, __func__);
+  MOZ_ASSERT(XRE_IsContentProcess());
+  // Only support PerformFinalChecks when we run ::CompleteAllowAccessFor in
+  // the same process. This callback is only used by eStorageAccessAPI,
+  // which is always runned in the same process.
+  MOZ_ASSERT(!aPerformFinalChecks);
+
+  ContentChild* cc = ContentChild::GetSingleton();
+  MOZ_ASSERT(cc);
+
+  return cc
+      ->SendCompleteAllowAccessFor(aParentContext, topLevelWindowId,
+                                   IPC::Principal(trackingPrincipal),
+                                   trackingOrigin, behavior, aReason)
+      ->Then(GetCurrentThreadSerialEventTarget(), __func__,
+             [](const ContentChild::CompleteAllowAccessForPromise::
+                    ResolveOrRejectValue& aValue) {
+               if (aValue.IsResolve() && aValue.ResolveValue().isSome()) {
+                 return StorageAccessGrantPromise::CreateAndResolve(
+                     aValue.ResolveValue().value(), __func__);
+               }
+               return StorageAccessGrantPromise::CreateAndReject(false,
+                                                                 __func__);
+             });
 }
 
 // CompleteAllowAccessFor is used to process the remaining work in
@@ -462,7 +484,7 @@
                  [](ParentAccessGrantPromise::ResolveOrRejectValue&& aValue) {
                    if (aValue.IsResolve()) {
                      return StorageAccessGrantPromise::CreateAndResolve(
-                         eAllow, __func__);
+                         ContentBlocking::eAllow, __func__);
                    }
                    return StorageAccessGrantPromise::CreateAndReject(false,
                                                                      __func__);